### PR TITLE
 Added Table View Panel

### PR DESCRIPTION
**Pull request by craftytrickster**
# Table View

**Summary**
Added new table panel that allows for the rendering of timeseries data in a table. 

screen shot 2015-04-27 at 12 11 25 pm

Now with improved pager! Exciting times!
screen shot 2015-05-18 at 11 32 14 am
# 

**Features**

screen shot 2015-05-19 at 3 32 32 pm

**Max Column Width:**  User can set rendered column width
**Sorting:**  Table can be sorted based on multiple columns, in both ascending and descending order (ex: ascending on column 1 and descending on column 2)
 **Paging:** User can define amount of results to be returned per table page
**Decimal Formatting:** User can specify number of decimals to display for numeric values
**Time as Date:** User can choose to see date either as a human readable date string or a unix timestamp
**Column Coloring:** User can specify thresholds (like in the single stat panel) in order to apply color highlighting to columns. Each column will have its own coloring options.
# 

(For strange data related reasons, some screenshots had to be taken with my phone instead of screen captured)

**Timeseries Table**

timeseries

As seen above, the timeseries table has a leftmost column of either Unix Timestamps or Date Strings, followed by a column per target query. This particular case has the column highlighting enabled. This should work will all datasources.
# 

**RAG Table**
rage

The RAG table allows you to view the state of a selected series of items. As seen here it has three different thresholds being hit. The RAG table requires one query per table row as per current limitations in the way datasource plugins work (can be expanded in future release)

screen shot 2015-05-19 at 3 30 02 pm

The RAG table also allows for the renaming of the column headers and it also allows the user to click on a row in order to be taken to a scripted dashboard hyperlink.

As seen in [issue 215](https://github.com/grafana/grafana/issues/215)

### CODE REVIEW

**Comment by cemo** [comment]
wuav :)

**Comment by hoffoo** [comment]
+1 would love to have this

**Comment by biancalana** [comment]
+1

**Comment by craftytrickster** [comment]
Grafana's datasources (at least for influxdbs) are meant to only return single key/value pairs. Currently, this is returned as [Timestamp/Value]. This is impeding the usefulness of the table widget, so I am making minimal changes in this branch to add flexibility, so that it returns [Timestamp/FirstValue, SecondValue.....]. The graphs will continue as before, since they only need the first value, but tables can take advantage of additional returned values.

**Comment by craftytrickster** [comment]
Update:
Within the next two weeks I plan to add RAG color highlighting functionality (like in the stat panel) to the table. Additionally, I think that I will actually add two types of tables:

Timeseries Table is what is currently seen in the attached image above.
However, I also will add status table, an example of a data center monitor would for the table would have the following format:
*DataCenters* | *State*
Melville | Green
Ronkonkoma | Yellow
Long Beach | Green
Great Neck | Red


PS: I am not sure why this is failing on appveyor, since I have not seen any issues locally and all the tests pass.

**Comment by torkelo** [comment]
its probably because you have not merged with master

**Comment by craftytrickster** [comment]
Considering that I was able to merge to/from master without any issues whatsoever, I am not sure why the following message is appearing: "continuous-integration/appveyor — AppVeyor was unable to build non-mergeable pull rer....."

File: public/app/components/settings.js [hunk 1]
```diff
@@ -19,6 +19,8 @@ function (_) {
         'singlestat': { path: 'panels/singlestat', name: 'Single stat' },
         'text':       { path: 'panels/text',       name: 'Text' },
         'dashlist':   { path: 'panels/dashlist',   name: 'Dashboard list' },
+        'table-time':   { path: 'panels/table/timeseries',   name: 'Timeseries Table' },
```
Comment (torkelo) [inline public/app/components/settings.js]: why are these defined here? are they used outside the table panel?

Comment (torkelo) [inline public/app/components/settings.js]: Sorry, ignore that comment, read it wrong (though it was the require config file)

**Comment by craftytrickster** [comment]
Hello,

Is there any possibility that this will be merged into the grafana repository in the near future? If so, in order to help, is there anything that I should change/add (other than merging in latest master) to facilitate the process?

**Comment by torkelo** [comment]
I will try to get test this and do a code review in the next couple of days

**Comment by craftytrickster** [comment]
Hi Torkel,

There are some lingering thoughts I have about the way grafana queries data that I wanted to share.

Let's say you have a graph panel with one query. Since you are plotting a single point on a graph at a specific time, the datasource plugins (at least influx) return data in [**TIMESTAMP** | **VALUE**] pairs. 

In other words, the app only supports single column queries. 

```SQL
SELECT MEAN(value) FROM .... 
```
And not this

```SQL
SELECT MEAN(value), COUNT(value), MAX(value) FROM .... 
```

As a result, in order to construct a table with many columns, the user will have to actually add a new query for every column to be returned. This does not seem like an ideal solution (although is fine as a starting point), since ideally the user would be able to combine multiple columns into one single request. Just wanted to point it out to you since it might make sense to make the plugins more flexible going forward to return data as such [**TIMESTAMP**| **ARRAY OF VALUES**] 

Supporting multiple column selection may require the query builder uis to be modified (unless you only allow it in raw query mode), since they seem to be wired for only single column selection

**Comment by torkelo** [comment]
InfluxDB 0.9 will only support single value column, but I guess if I want to support Elasticsearch queries (for the table panel), and SQL queries, I think you are right. but not sure, think another option is to to types of queries for the datasource abstraction, one that returns single series time series data, another that returns untyped table data.  

But need to think about this more.

**Comment by torkelo** [comment]
Had time to test this right now. Feel that it needs some more work but a really good start! 

Would like to merge the to table panels into one, and have a switch for instead to control if the table panel should show the values or series summaries. 

I feel that date format should be enabled by default, and also maybe a an option to specify how the dates should be formatted. 
I could not get the column width option to work, think the table should be width: 100% by default so it fill the panel. 

Also the results per page should probably be calculated automatically so there isnt overflow or if there is overflow then a scrollbar maybe. 

image

**Comment by torkelo** [comment]
Also the paging should probably say how  many pages there are.

**Comment by craftytrickster** [comment]
I added new a new LESS file to the project, which has the overflow protection/formatting issues you are discussing above. If you run `grunt css` it should fix the overflow as seen in the screenshot. 

As for the paging, that is something that is already precalculated, I just couldn't figure out a good way to display it. I can add it in for you though in the next few days. Also, I can set the human readable date to default as you suggested.

**Comment by craftytrickster** [comment]
Merged the tables into one table type. Before testing `grunt css` is required to get the tablePanel.less contents.

**Comment by craftytrickster** [comment]
After discussing RAG table usage with my colleague, I realized that my approach (forcing use of InfluxDb9) is completely unnecessary. I will simplify RAG to work with normal queries in the next day or two.

**Comment by torkelo** [comment]
nice

**Comment by strangereous** [comment]
hi, do you have an ETA on the release for the table feature? It looks really cool, really want to try it out :+1:

**Comment by craftytrickster** [comment]
I'm glad you like it. I have no ETA, it really depends on when Torkel wants to accept it into the branch. Additionally he might want me to implement some changes before that happens. If you want to try it out now, it does fully function with this branch. Just be sure to use grunt css first, since I added LESS data.

**Comment by biancalana** [comment]
that's look a great feature !

**Comment by aldanor** [comment]
Looks good! That would be very useful

**Comment by sathibault** [comment]
A table panel would be great.  Would there be a way to use this to display textual data?  I'm thinking of something like a leader board.

**Comment by steverweber** [comment]
text data would also be extremely helpful to display log file messages!

**Comment by craftytrickster** [comment]
The table itself can render any kind of data (including text) without any issue. 

However, the table is bound by the current query builders, which typically are designed for graphical timestamp data, so that is something to keep in mind. However, if you wanted to display log messages against timestamps, that would work fine.

**Comment by torkelo** [comment]
I am really sorry that this hasnt been merged yet. Hope to find time for this at the start of 2.2 milestone

**Comment by steverweber** [comment]
Is it to tricky to merge in for milestone 2.1?  I don't mind waiting a day extra for it :) This is a killer feature!

**Comment by torkelo** [comment]
It needs a couple of days work I think

**Comment by steverweber** [comment]
I'm a little biased because I need this table feature for a project that is due in a few weeks.  I would be vary grateful if you held back the 2.1 milestone for this merge... But yap its your call :+1:

**Comment by Pricey3000** [comment]
This table feature would be extremely useful for a project I'm running too so I second steverweber's comment :+1:

**Comment by steverweber** [comment]
how about adding this guy to milestone 2.2

**Comment by The-Nik** [comment]
+1 for 2.2

**Comment by steverweber** [comment]
+ for 2.5

**Comment by timgriffiths** [comment]
+1

**Comment by sysadminmike** [comment]
This is great - im hooking stuff up to postgres as the metrics backend and for logs this is perfect - is it possible to add a filter field so you can enter a text string like the $timeFilter is passed to the metric query?   

So its then possible to search/filter the logs/table data with sql or for people using elasticsearch for metrics (when backend exists)

**Comment by torkelo** [comment]
closed in favor of https://github.com/grafana/grafana/issues/3219

**Pull request closed without merging**